php_phongo.c

```diff
@@ -1222,6 +1262,38 @@ bool php_phongo_server_to_zval(zval* retval, mongoc_server_description_t* sd) /*
 	return true;
 } /* }}} */
 
+bool php_phongo_topology_description_to_zval(zval* retval, mongoc_topology_description_t* td) /* {{{ */
+{
+	array_init(retval);
+
+	{
+		zval                          servers;
+		size_t                        i, n = 0;
+		mongoc_server_description_t** sds = mongoc_topology_description_get_servers(td, &n);
+
+		array_init_size(&servers, n);
+
+		for (i = 0; i < n; i++) {
+			zval obj;
+
+			if (!php_phongo_server_description_to_zval(&obj, sds[i])) {
+				/* Exception already thrown */
+				zval_ptr_dtor(&obj);
+				zval_ptr_dtor(&servers);
+				mongoc_server_descriptions_destroy_all(sds, n);
```

Comment:
Is `mongoc_server_descriptions_destroy_all` part of the general cleanup done when recovering from an error, why isn't is also called during the non-error code path? I believe this leaks; however, it's hidden by the fact that the TopologyDescription debug test currently dumps a description with no servers (i.e. `n=0`).

Changing `topologyDescription-debug-001.phpt` to dump the final TopologyDescription results in the following in `topologyDescription-debug-001.mem` when running with `TEST_PHP_ARGS=-m` (Valgrind):

```
==1239422== 8 bytes in 1 blocks are indirectly lost in loss record 5 of 33
==1239422==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==1239422==    by 0x784468: __zend_malloc (zend_alloc.c:3030)
==1239422==    by 0x7844D5: __zend_calloc (zend_alloc.c:3042)
==1239422==    by 0x8379BE6: php_phongo_calloc (php_phongo.c:3434)
==1239422==    by 0x82B1B92: bson_malloc0 (bson-memory.c:108)
==1239422==    by 0x82FD6AE: mongoc_generation_map_new (mongoc-generation-map.c:61)
==1239422==    by 0x82FD6D2: mongoc_generation_map_copy (mongoc-generation-map.c:71)
==1239422==    by 0x8323991: mongoc_server_description_new_copy (mongoc-server-description.c:830)
==1239422==    by 0x833BA73: mongoc_topology_description_get_servers (mongoc-topology-description.c:2281)
==1239422==    by 0x83A8581: php_phongo_topologydescription_get_properties_hash (TopologyDescription.c:172)
==1239422==    by 0x83A8705: php_phongo_topologydescription_get_debug_info (TopologyDescription.c:208)
==1239422==    by 0x8A0A6A: zend_std_get_properties_for (zend_object_handlers.c:1779)
==1239422== 
==1239422== 64 bytes in 1 blocks are indirectly lost in loss record 22 of 33
==1239422==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==1239422==    by 0x784468: __zend_malloc (zend_alloc.c:3030)
==1239422==    by 0x8379BBD: php_phongo_malloc (php_phongo.c:3429)
==1239422==    by 0x82B1B11: bson_malloc (bson-memory.c:70)
==1239422==    by 0x829FA85: bson_copy_to (bson.c:2222)
==1239422==    by 0x8324A0F: mongoc_server_description_set_topology_version (mongoc-server-description.c:1252)
==1239422==    by 0x83234F8: mongoc_server_description_handle_hello (mongoc-server-description.c:728)
==1239422==    by 0x8323923: mongoc_server_description_new_copy (mongoc-server-description.c:814)
==1239422==    by 0x833BA73: mongoc_topology_description_get_servers (mongoc-topology-description.c:2281)
==1239422==    by 0x83A8581: php_phongo_topologydescription_get_properties_hash (TopologyDescription.c:172)
==1239422==    by 0x83A8705: php_phongo_topologydescription_get_debug_info (TopologyDescription.c:208)
==1239422==    by 0x8A0A6A: zend_std_get_properties_for (zend_object_handlers.c:1779)
==1239422== 
==1239422== 1,024 bytes in 1 blocks are indirectly lost in loss record 29 of 33
==1239422==    at 0x483DFAF: realloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==1239422==    by 0x78451B: __zend_realloc (zend_alloc.c:3049)
==1239422==    by 0x8379C0F: php_phongo_realloc (php_phongo.c:3439)
==1239422==    by 0x82B1C28: bson_realloc (bson-memory.c:154)
==1239422==    by 0x82B1C97: bson_realloc_ctx (bson-memory.c:194)
==1239422==    by 0x829AF1F: _bson_impl_alloc_grow (bson.c:188)
==1239422==    by 0x829AF8A: _bson_grow (bson.c:222)
==1239422==    by 0x829B0CC: _bson_append_va (bson.c:326)
==1239422==    by 0x829B43B: _bson_append (bson.c:412)
==1239422==    by 0x829CCA5: bson_append_int32 (bson.c:1200)
==1239422==    by 0x829D6AB: bson_append_iter (bson.c:1399)
==1239422==    by 0x829FC28: bson_copy_to_excluding_noinit_va (bson.c:2263)
==1239422== 
==1239422== 2,144 bytes in 1 blocks are indirectly lost in loss record 31 of 33
==1239422==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==1239422==    by 0x784468: __zend_malloc (zend_alloc.c:3030)
==1239422==    by 0x7844D5: __zend_calloc (zend_alloc.c:3042)
==1239422==    by 0x8379BE6: php_phongo_calloc (php_phongo.c:3434)
==1239422==    by 0x82B1B92: bson_malloc0 (bson-memory.c:108)
==1239422==    by 0x83237D1: mongoc_server_description_new_copy (mongoc-server-description.c:795)
==1239422==    by 0x833BA73: mongoc_topology_description_get_servers (mongoc-topology-description.c:2281)
==1239422==    by 0x83A8581: php_phongo_topologydescription_get_properties_hash (TopologyDescription.c:172)
==1239422==    by 0x83A8705: php_phongo_topologydescription_get_debug_info (TopologyDescription.c:208)
==1239422==    by 0x8A0A6A: zend_std_get_properties_for (zend_object_handlers.c:1779)
==1239422==    by 0x8A0B57: zend_get_properties_for (zend_object_handlers.c:1808)
==1239422==    by 0x6CBC89: php_var_dump (var.c:154)
==1239422== 
==1239422== 3,248 (8 direct, 3,240 indirect) bytes in 1 blocks are definitely lost in loss record 33 of 33
==1239422==    at 0x483B7F3: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==1239422==    by 0x784468: __zend_malloc (zend_alloc.c:3030)
==1239422==    by 0x7844D5: __zend_calloc (zend_alloc.c:3042)
==1239422==    by 0x8379BE6: php_phongo_calloc (php_phongo.c:3434)
==1239422==    by 0x82B1B92: bson_malloc0 (bson-memory.c:108)
==1239422==    by 0x833BA11: mongoc_topology_description_get_servers (mongoc-topology-description.c:2272)
==1239422==    by 0x83A8581: php_phongo_topologydescription_get_properties_hash (TopologyDescription.c:172)
==1239422==    by 0x83A8705: php_phongo_topologydescription_get_debug_info (TopologyDescription.c:208)
==1239422==    by 0x8A0A6A: zend_std_get_properties_for (zend_object_handlers.c:1779)
==1239422==    by 0x8A0B57: zend_get_properties_for (zend_object_handlers.c:1808)
==1239422==    by 0x6CBC89: php_var_dump (var.c:154)
==1239422==    by 0x6CC192: zif_var_dump (var.c:218)
==1239422== 
```

Note: Unlike the other leak I mentioned, PHP doesn't report _this_ leak itself since it exists entirely in libmongoc. Leaks reported by PHP's debug mode are mainly with zval usage and internal PHP memory.